ipa-core/src/protocol/ipa_prf/boolean_ops/multiplication.rs

```diff
@@ -1,33 +1,164 @@
+        test_executor::run,
+        test_fixture::{Reconstruct, Runner, TestWorld},
+    };
+
+    fn as_i128<B>(x: B) -> i128
+    where
+        B: BooleanArray + U128Conversions,
+    {
+        let mut out: i128 = i128::try_from(x.as_u128()).unwrap();
+        let msb = (out >> (B::BITS - 1)) & 1;
+        out -= msb * (1 << B::BITS);
+        out
+    }
+
+    #[test]
+    #[allow(clippy::cast_precision_loss)]
+    fn semi_honest_mul() {
+        run(|| async move {
+            let world = TestWorld::default();
+
+            let mut rng = thread_rng();
+
+            let all_x_values = (0..256)
+                .map(|i| BA8::truncate_from(u128::try_from(i).unwrap()))
+                .collect::<Vec<_>>();
+            let random_y_values = (0..256).map(|_| rng.gen::<BA8>()).collect::<Vec<_>>();
+
+            let result: Vec<BA16> = world
+                .upgraded_semi_honest(
+                    all_x_values
+                        .clone()
+                        .into_iter()
+                        .zip(random_y_values.clone()),
+                    |ctx, x_y_vals| async move {
+                        let (x_vals, y_vals): (Vec<AdditiveShare<BA8>>, Vec<AdditiveShare<BA8>>) =
+                            x_y_vals.into_iter().unzip();
+                        let mut vectorized_x_inputs: BitDecomposed<AdditiveShare<Boolean, 256>> =
+                            BitDecomposed::new(iter::empty());
+                        let _ = vectorized_x_inputs.transpose_from(&x_vals);
+
+                        let mut vectorized_y_inputs: BitDecomposed<AdditiveShare<Boolean, 256>> =
+                            BitDecomposed::new(iter::empty());
+                        let _ = vectorized_y_inputs.transpose_from(&y_vals);
+
+                        let result = integer_mul::<_, DefaultBitStep, 256>(
+                            ctx.set_total_records(1),
+                            RecordId::FIRST,
+                            &vectorized_x_inputs,
+                            &vectorized_y_inputs,
+                        )
+                        .await
+                        .unwrap();
+
+                        Vec::transposed_from(&result).unwrap()
+                    },
+                )
+                .await
+                .reconstruct();
+
+            for ((res, x), y) in result
+                .iter()
+                .zip(all_x_values.iter())
+                .zip(random_y_values.iter())
+            {
+                let y_as_signed_number: i128 = as_i128(*y);
+                let x_as_unsigned_number: i128 = x.as_u128() as i128;
+                let expected: i128 = y_as_signed_number * x_as_unsigned_number;
+
+                assert_eq!(
+                    (x, y, y_as_signed_number, x_as_unsigned_number, expected),
+                    // (x, y, y_as_unsigned_number, x_as_signed_number, res.as_u128())
+                    (x, y, y_as_signed_number, x_as_unsigned_number, as_i128(*res))
```

Comment:
Do we really need x, y, y_as_signed_number and x_as_unsigned_number? Was this just for debugging? Maybe we can just use x and y in a "signed" format so that the error message is easy to read.